core/sizes.cpp

```diff
@@ -2,81 +2,175 @@
 /// @file
 /// @author Simon Heybrock
-#include "scipp/core/sizes.h"
+#include <algorithm>
+
 #include "scipp/core/except.h"
+#include "scipp/core/sizes.h"
 
 namespace scipp::core {
 
-Sizes::Sizes(const Dimensions &dims) {
-  for (const auto dim : dims.labels())
-    m_sizes[dim] = dims[dim];
+namespace {
+template <class T> void expectUnique(const T &map, const Dim label) {
+  if (map.contains(label))
+    throw except::DimensionError("Duplicate dimension.");
 }
 
-bool Sizes::operator==(const Sizes &other) const {
-  return m_sizes == other.m_sizes;
+template <class T> void expectExtendable(const T &map) {
+  if (map.size() == T::capacity)
+    throw except::DimensionError(
+        "Maximum number of allowed dimensions exceeded.");
 }
 
-bool Sizes::operator!=(const Sizes &other) const { return !operator==(other); }
+template <class T> std::string to_string(const T &map) {
+  std::string repr("[");
+  for (const auto &key : map)
+    repr += to_string(key) + ":" + std::to_string(map[key]) + ", ";
+  repr += "]";
+  return repr;
+}
 
-void Sizes::clear() { m_sizes.clear(); }
+template <class T>
+void throw_dimension_not_found_error(const T &expected, Dim actual) {
+  throw except::DimensionError{"Expected dimension to be in " +
+                               to_string(expected) + ", got " +
+                               to_string(actual) + '.'};
+}
 
-scipp::index Sizes::operator[](const Dim dim) const { return at(dim); }
+} // namespace
 
-scipp::index &Sizes::operator[](const Dim dim) { return at(dim); }
+template <class Key, class Value, int16_t Capacity>
+bool small_stable_map<Key, Value, Capacity>::operator==(
+    const small_stable_map &other) const noexcept {
+  if (size() != other.size())
+    return false;
+  for (const auto &key : *this)
+    if (!other.contains(key) || at(key) != other.at(key))
+      return false;
+  return true;
+}
+
+template <class Key, class Value, int16_t Capacity>
+bool small_stable_map<Key, Value, Capacity>::operator!=(
+    const small_stable_map &other) const noexcept {
+  return !operator==(other);
+}
+
+template <class Key, class Value, int16_t Capacity>
+typename std::array<Key, Capacity>::const_iterator
+small_stable_map<Key, Value, Capacity>::find(const Key &key) const {
+  return std::find(begin(), end(), key);
+}
 
-scipp::index Sizes::at(const Dim dim) const {
-  scipp::expect::contains(*this, dim);
-  return m_sizes.at(dim);
+template <class Key, class Value, int16_t Capacity>
+bool small_stable_map<Key, Value, Capacity>::contains(const Key &key) const {
+  return find(key) != end();
 }
 
-scipp::index &Sizes::at(const Dim dim) {
-  scipp::expect::contains(*this, dim);
-  return m_sizes.at(dim);
+template <class Key, class Value, int16_t Capacity>
+scipp::index
+small_stable_map<Key, Value, Capacity>::index(const Key &key) const {
+  const auto it = find(key);
+  if (it == end())
+    throw_dimension_not_found_error(*this, key);
```

Comment:
This is already supposing the type for `Key`?

Reply:
Might be, but as it is all other exceptions also throw errors specific to dimensions, so this would just be one of the the things to fix if we ever use this map for another purpose.